problems/src/word_ladder/step2.rs

```diff
@@ -0,0 +1,125 @@
+    双方向BFSで、ならしだが探索時間が短くできる
+
+*/
+
+pub struct Solution;
+
+type Step = i32;
+
+use std::collections::{HashMap, HashSet, VecDeque};
+
+impl Solution {
+    fn visit_nodes<'a>(
+        queue: &mut VecDeque<(&'a str, Step)>,
+        visited: &mut HashMap<&'a str, Step>,
+        other_visited: &HashMap<&'a str, Step>,
+        intermediate_word_map: &HashMap<String, Vec<&'a str>>,
+    ) -> Option<Step> {
+        for _ in 0..queue.len() {
+            let (word, step) = queue.pop_front().unwrap();
+            for i in 0..word.len() {
+                let intermediate = format!("{}*{}", &word[0..i], &word[i + 1..]);
+                for next_word in intermediate_word_map.get(&intermediate).unwrap_or(&vec![]) {
+                    if other_visited.contains_key(*next_word) {
+                        return Some(step + other_visited[next_word]);
+                    }
+                    if !visited.contains_key(*next_word) {
+                        visited.insert(&next_word, step + 1);
+                        queue.push_back((next_word, step + 1));
+                    }
+                }
+            }
+        }
+
+        None
+    }
+
+    pub fn ladder_length(begin_word: String, end_word: String, word_list: Vec<String>) -> i32 {
+        if !word_list.contains(&end_word) {
+            return 0;
+        }
+
+        let mut intermediate_word_map = HashMap::<String, Vec<&str>>::new();
+
+        for word in word_list.iter() {
+            for i in 0..word.len() {
+                let intermediate = format!("{}*{}", &word[0..i], &word[i + 1..]);
+                intermediate_word_map
+                    .entry(intermediate.to_string())
+                    .or_insert(vec![])
+                    .push(&word);
+            }
+        }
+
+        let mut queue_begin = VecDeque::<(&str, Step)>::new();
+        let mut visited_begin = HashMap::<&str, Step>::new();
+        let mut queue_end = VecDeque::<(&str, Step)>::new();
+        let mut visited_end = HashMap::<&str, Step>::new();
+
+        queue_begin.push_back((begin_word.as_str(), 1));
+        visited_begin.insert(&begin_word, 1);
+        queue_end.push_back((end_word.as_str(), 1));
+        visited_end.insert(&end_word, 1);
+
+        // bidirectional breadth first search
+        while !queue_begin.is_empty() && !queue_end.is_empty() {
+            // queueの小さい方から探索してメモリ使用量を抑える
+            if let Some(steps) = if queue_begin.len() <= queue_end.len() {
+                Self::visit_nodes(
+                    &mut queue_begin,
+                    &mut visited_begin,
+                    &mut visited_end,
+                    &intermediate_word_map,
+                )
+            } else {
+                Self::visit_nodes(
+                    &mut queue_end,
+                    &mut visited_end,
+                    &mut visited_begin,
+                    &intermediate_word_map,
+                )
+            } {
+                return steps;
+            }
```

Comment:
分岐させた計算の返り値をif letでパターンマッチさせてるのちょっと複雑な気がしました。
↓は名前が微妙ですが、別々に分けた方がいい気がします。

```suggestion
            let step_opt = if queue_begin.len() <= queue_end.len() {
                Self::visit_nodes(
                    &mut queue_begin,
                    &mut visited_begin,
                    &mut visited_end,
                    &intermediate_word_map,
                )
            } else {
                Self::visit_nodes(
                    &mut queue_end,
                    &mut visited_end,
                    &mut visited_begin,
                    &intermediate_word_map,
                )
            };
            if let Some(step) = step_opt {
                return step;
            }
```

Reply:
ありがとうございます
たしかにこれ分かりにくくなってますね

Reply:
@fhiyo 
[522a077](https://github.com/Yoshiki-Iwasa/Arai60/pull/22/commits/522a077982a2694a9e146669c5c4931e093b2f99)

ここで同じコードを繰り返すのを避けるために、`queue`と`visited`を一つにしました。
読みやすや的にどうでしょうか？？

Reply:
読みやすいと思います 👍